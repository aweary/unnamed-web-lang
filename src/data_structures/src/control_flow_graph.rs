use petgraph::dot::Dot;
use petgraph::stable_graph::{NodeIndex, StableGraph};

use std::collections::VecDeque;
use std::fmt::{self, Debug};

#[derive(Debug)]
struct IdentDistribution;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct BlockIndex(NodeIndex);


pub trait Blockable : Debug {
    fn has_early_exit(&self) -> bool;
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct PartialFromEdge(pub BlockIndex, pub ControlFlowEdge);

// struct PartialToEdge(BlockIndex, ControlFlowEdge);

/// A control flow graph for a single function
#[derive(Debug, Clone)]
pub struct ControlFlowGraph<T: Blockable> {
    /// The graph data structure
    graph: StableGraph<BasicBlock<T>, ControlFlowEdge>,
    /// The entry block of the graph
    /// TODO make private
    pub entry_block: BlockIndex,
    /// The exit block of the graph
    exit_block: BlockIndex,
    /// The first basic block block
    first_block: Option<BlockIndex>,
    /// The last block added to the graph
    last_block: Option<BlockIndex>,
    /// A queue of edges that will be flushed when a block is added
    edge_queue: VecDeque<(BlockIndex, ControlFlowEdge)>,
}

impl<T: Blockable> Default for ControlFlowGraph<T> {
    fn default() -> Self {
        let mut graph = StableGraph::default();
        // Initialize the special entry and exit blocks
        let entry_block = BlockIndex(graph.add_node(BasicBlock::Entry));
        let exit_block = BlockIndex(graph.add_node(BasicBlock::Exit));
        // Initialize the first basic block, since we know that
        // we will at the very least have one
        ControlFlowGraph {
            graph,
            entry_block,
            exit_block,
            first_block: None,
            last_block: None,
            edge_queue: VecDeque::default(),
        }
    }
}

impl<T : Blockable> ControlFlowGraph<T> {
    /// Add a new basic block to the grapg
    pub fn add_block(&mut self, block: Block<T>) -> BlockIndex {
        let block_index = BlockIndex(self.graph.add_node(BasicBlock::Block(block)));
        // If this is the first block to be added to the graph, add an edge from the
        // special entry block to this one.
        if self.first_block.is_none() {
            self.first_block = Some(block_index);
            self.add_edge(self.entry_block, block_index, ControlFlowEdge::Normal);
        }

        // Flush the edge
        for (from_block, edge) in self.edge_queue.clone() {
            self.add_edge(from_block, block_index, edge);
        }
        // Clear out the edge queue
        self.edge_queue.clear();
        self.last_block = Some(block_index);
        block_index
    }

    /// Get the last basic block added to the graph, or `None` if it's empty
    pub fn last_block(&self) -> Option<BlockIndex> {
        self.last_block
    }

    /// Add an edge from a block to the special exit block
    pub fn add_edge_to_exit(&mut self, index: BlockIndex, edge: ControlFlowEdge) {
        self.add_edge(index, self.exit_block, edge);
    }

    /// Add an edge between two blocks
    pub fn add_edge(&mut self, from: BlockIndex, to: BlockIndex, edge: ControlFlowEdge) {
        self.graph.update_edge(from.0, to.0, edge);
    }

    /// Queues up an edge for the next block added to the graph, where `block_index` is the
    /// source block.
    pub fn enqueue_edge(&mut self, block_index: BlockIndex, edge: ControlFlowEdge) {
        self.edge_queue.push_back((block_index, edge));
    }

    /// Returns whether a given block has early exit semantics. In other words, it tells
    /// us whether we need to account for the default top-down flow for non-early exiting
    /// statements.
    pub fn block_has_early_exit(&self, block_index: BlockIndex) -> bool {
        match &self.graph[block_index.0] {
            BasicBlock::Block(block) => {
                if let Some(stmt) = block.statements.last() {
                    stmt.has_early_exit()
                } else {
                    false
                }
            }
            _ => false,
        }
    }

    /// Flush the edge queue using an arbitrary block as the target
    pub fn flush_edge_queue(&mut self, block_index: BlockIndex) {
        // Flush the edge queue
        for (from_block, edge) in self.edge_queue.clone() {
            self.graph.update_edge(from_block.0, block_index.0, edge);
        }
        // Clear out the edge queue
        self.edge_queue.clear();
    }

    /// Takes all the queued edges and connects them with the special exit
    /// block. This signals that CFG construction is completed.
    /// We also check if the graph is empty, and connect the
    /// entry and exit blocks if so.
    pub fn flush_edge_queue_to_exit_block(&mut self) {
        self.flush_edge_queue(self.exit_block);
        // If there is no last block, nothing was ever added to this
        // graph.
        if self.last_block.is_none() {
            self.add_edge(self.entry_block, self.exit_block, ControlFlowEdge::Normal);
        }
    }

    pub fn print(&self) {
        println!("{:?}", Dot::with_config(&self.graph, &[]));
    }

    pub fn graphviz_output(&self) -> String {
        format!("{:?}", Dot::with_config(&self.graph, &[]))
    }
}

#[derive(Clone)]
pub struct Block<T: Blockable> {
    // The set of non-branching statements within the block.
    pub statements: Vec<T>,
}

impl<T: Blockable> Default for Block<T> {
    fn default() -> Self {
        Block { statements: vec![] }
    }
}

// Only print the number of statements in a block, makes debugging easier for now
impl<T: Blockable> fmt::Debug for Block<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self.statements)
    }
}

impl<T : Blockable> Block<T> {
    /// Push a statement into the basic block
    pub fn push(&mut self, stmt: T) {
        self.statements.push(stmt);
    }

    /// returns whether this block contains any statements
    pub fn is_empty(&self) -> bool {
        self.statements.is_empty()
    }

    // pub fn exit_type(&self) -> BlockExit {
    //     match self.statements.last() {
    //         Some(stmt) => {
    //             use crate::hir::StatementKind;
    //             match stmt.kind {
    //                 StatementKind::BranchingCondition => BlockExit::BranchCondition,
    //                 StatementKind::LoopingCondition => BlockExit::LoopCondition,
    //                 StatementKind::Return => BlockExit::Return,
    //                 _ => BlockExit::Normal,
    //             }
    //         }
    //         None => BlockExit::Normal,
    //     }
    // }
}

#[derive(Debug, Clone)]
pub enum BlockExit {
    // Placeholder, we don't know how the block exits yet
    Unknown,
    // The block forks based on a boolean condition. It will
    // have edges to other blocks for both `true` and `false`
    BranchCondition,
    LoopCondition,
    // The block ends with a return statement. It will have a
    // single edge to the special exit block.
    Return,
    // The block naturally flows to another block, e.g., when
    // an `if` block finishes executing.
    Normal,
}

#[derive(Debug, Clone)]
pub enum BasicBlock<T : Blockable> {
    /// A special block denoting the entry-point of the graph.
    Entry,
    /// A basic block.
    Block(Block<T>),
    // Special block denoting the exit-point of the graph.
    Exit,
}

/// An edge between basic blocks. It answers the question
/// "how can control flow move between blocks?"
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ControlFlowEdge {
    /// Normal control flow is what occurs as code is executed
    /// top-down. For example, when an `if` without an `else` clause
    /// ends it naturally moves control flow back outside the block.
    Normal,
    /// When control flow depends on some boolean condition, this is
    /// the page taken when the condition evalutes to `true`.
    ConditionTrue,
    /// Same as ConditionTrue, but for `false`
    ConditionFalse,
    /// An explicit return statement, always goes straight to the exit block.
    Return,
}
