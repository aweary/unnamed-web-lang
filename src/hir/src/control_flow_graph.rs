use petgraph::dot::Dot;
use petgraph::stable_graph::{NodeIndex, StableGraph};
use petgraph::visit::Dfs;

// use petgraph::Direction;
// use petgraph::algo::{all_simple_paths, astar, dominators, has_path_connecting};

use std::collections::VecDeque;
use std::fmt;

use quickcheck::{Arbitrary, Gen, TestResult, Testable};


use rand::distributions::{Distribution};
use rand::Rng;

const MAX_MOCK_CFG_CODE_SNIPPET_DEPTH : usize = 500;

#[derive(Debug)]
struct IdentDistribution;

impl Distribution<char> for IdentDistribution {
    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> char {
        const RANGE: u32 = 26 + 26 + 2;
        const VALID_ASCII_IDENT_CHARS: &[u8] =
            b"ABCDEFGHIJKLMNOPQRSTUVWXYZ\
                abcdefghijklmnopqrstuvwxyz\
                $_";
        loop {
            let var = rng.next_u32() >> (32 - 6);
            if var < RANGE {
                return VALID_ASCII_IDENT_CHARS[var as usize] as char
            }
        }
    }
}


/// Generate a random snippet of raw code that should
/// always give us a well-formed control-flow-graph.
#[derive(Default, Clone, Debug)]
pub struct MockControlFlowCodeSnippetBuilder {
    snippet: String,
    depth: usize,
}

impl MockControlFlowCodeSnippetBuilder {
    pub fn code(&self) -> &str {
        &self.snippet
    }

    pub fn build(&mut self) {
        self.build_block();
    }

    fn push(&mut self, code: &str) {
        // for _ in 0..self.depth {
        //     self.snippet.push(' ');
        // }
        self.snippet.push_str(code);
    }

    fn build_block(&mut self) {
        self.depth += 1;
        self.push("{\n");
        // let block_size : usize = rand::random();
        let block_size = 5;
        for _ in 0..block_size {
            self.build_stmt();
        }
        if rand::random() {
        use rand::{thread_rng, Rng};
            // Add a return statement
        let value: String = thread_rng().sample_iter(&IdentDistribution).take(5).collect();   
        self.push(&format!("return \"{}\";", value));
        }
        self.push("\n}");
        self.depth -= 1;
    }

    fn build_stmt(&mut self) {
        if self.snippet.len() >= MAX_MOCK_CFG_CODE_SNIPPET_DEPTH {
            return;
        }
        // Eeach statement has equal random chance
        if rand::random() {
            self.build_non_branching_stmt();
        } else if rand::random() {
            self.build_if_branch();
        } else if rand::random() {
            self.build_while_loop();
        }
    }

    fn build_non_branching_stmt(&mut self) {
        use rand::distributions::Alphanumeric;
        use rand::{thread_rng, Rng};
        let name: String = thread_rng().sample_iter(&IdentDistribution).take(5).collect();
        let value: String = thread_rng().sample_iter(&IdentDistribution).take(5).collect();
        let code = format!("let {} = \"{}\";\n", name, value);
        self.push(&code);
    }

    fn build_if_branch(&mut self) {
        self.push("if true ");
        self.build_block();
        // Else-If chain
        if rand::random() {
            // self.push(" else ");
        }
        // Else block
        else if rand::random() {
            self.push(" else ");
            self.build_block();
        }
        self.push(";");
    }

    fn build_while_loop(&mut self) {
        self.push("while true ");
        self.build_block();
        self.push(";");
    }
}

impl Arbitrary for MockControlFlowCodeSnippetBuilder {
    fn arbitrary<G: Gen>(g: &mut G) -> Self {
        let mut builder = MockControlFlowCodeSnippetBuilder::default();
        builder.build();
        builder
    }
}

impl Arbitrary for ControlFlowGraph {
    fn arbitrary<G: Gen>(g: &mut G) -> Self {
        /// Start with an empty graph
        let mut cfg = ControlFlowGraph::default();
        // ...
        cfg
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct BlockIndex(NodeIndex);

use crate::hir;

pub struct SubControlFlowGraph<'cfg> {
    cfg: &'cfg mut ControlFlowGraph,
}

/// A control flow graph for a single function
#[derive(Debug, Clone)]
pub struct ControlFlowGraph {
    /// The graph data structure
    graph: StableGraph<BasicBlock, ControlFlowEdge>,
    /// The entry block of the graph
    /// TODO make private
    pub entry_block: BlockIndex,
    /// The exit block of the graph
    exit_block: BlockIndex,
    /// The first basic block block
    first_block: Option<BlockIndex>,
    /// The last block added to the graph
    last_block: Option<BlockIndex>,
    /// A queue of edges that will be flushed when a block is added
    edge_queue: VecDeque<(BlockIndex, ControlFlowEdge)>,
}

impl Default for ControlFlowGraph {
    fn default() -> Self {
        let mut graph = StableGraph::default();
        // Initialize the special entry and exit blocks
        let entry_block = BlockIndex(graph.add_node(BasicBlock::Entry));
        let exit_block = BlockIndex(graph.add_node(BasicBlock::Exit));
        // Initialize the first basic block, since we know that
        // we will at the very least have one
        ControlFlowGraph {
            graph,
            entry_block,
            exit_block,
            first_block: None,
            last_block: None,
            edge_queue: VecDeque::default(),
        }
    }
}

impl ControlFlowGraph {
    /// Add a new basic block to the grapg
    pub fn add_block(&mut self, block: Block) -> BlockIndex {
        let block_index = BlockIndex(self.graph.add_node(BasicBlock::Block(block)));
        // If this is the first block to be added to the graph, add an edge from the
        // special entry block to this one.
        if self.first_block.is_none() {
            self.first_block = Some(block_index);
            self.add_edge(self.entry_block, block_index, ControlFlowEdge::Normal);
        }

        // Flush the edge
        for (from_block, edge) in self.edge_queue.clone() {
            self.add_edge(from_block, block_index, edge);
        }
        // Clear out the edge queue
        self.edge_queue.clear();
        self.last_block = Some(block_index);
        block_index
    }

    pub fn dfs(&self) {
        let mut dfs = Dfs::new(&self.graph, self.entry_block.0);
        while let Some(index) = dfs.next(&self.graph) {
            println!("node {:?}", index);
        }
    }

    /// Get the last basic block added to the graph, or `None` if it's empty
    pub fn last_block(&self) -> Option<BlockIndex> {
        self.last_block
    }

    /// Add an edge from a block to the special exit block
    pub fn add_edge_to_exit(&mut self, index: BlockIndex, edge: ControlFlowEdge) {
        self.add_edge(index, self.exit_block, edge);
    }

    /// Add an edge between two blocks
    pub fn add_edge(&mut self, from: BlockIndex, to: BlockIndex, edge: ControlFlowEdge) {
        self.graph.update_edge(from.0, to.0, edge);
    }

    /// Queues up an edge for the next block added to the graph, where `block_index` is the
    /// source block.
    pub fn enqueue_edge(&mut self, block_index: BlockIndex, edge: ControlFlowEdge) {
        self.edge_queue.push_back((block_index, edge));
    }

    /// Returns whether a given block has early exit semantics. In other words, it tells
    /// us whether we need to account for the default top-down flow for non-early exiting
    /// statements.
    pub fn block_has_early_exit(&self, block_index: BlockIndex) -> bool {
        match &self.graph[block_index.0] {
            BasicBlock::Block(block) => {
                if let Some(stmt) = block.statements.last() {
                    use crate::hir::StatementKind::*;
                    match stmt.kind {
                        Return => true,
                        _ => false,
                    }
                } else {
                    false
                }
            }
            _ => false,
        }
    }

    /// Flush the edge queue using an arbitrary block as the target
    pub fn flush_edge_queue(&mut self, block_index: BlockIndex) {
        // Flush the edge queue
        for (from_block, edge) in self.edge_queue.clone() {
            self.graph.update_edge(from_block.0, block_index.0, edge);
        }
        // Clear out the edge queue
        self.edge_queue.clear();
    }

    /// Takes all the queued edges and connects them with the special exit
    /// block. This signals that CFG construction is completed.
    /// We also check if the graph is empty, and connect the
    /// entry and exit blocks if so.
    pub fn flush_edge_queue_to_exit_block(&mut self) {
        self.flush_edge_queue(self.exit_block);
        // If there is no last block, nothing was ever added to this
        // graph.
        if self.last_block.is_none() {
            self.add_edge(self.entry_block, self.exit_block, ControlFlowEdge::Normal);
        }
    }

    pub fn print(&self) {
        println!("{:?}", Dot::with_config(&self.graph, &[]));
    }

    pub fn graphviz_output(&self) -> String {
        format!("{:?}", Dot::with_config(&self.graph, &[]))
    }

    // ...
}

#[derive(Default, Clone)]
pub struct Block {
    // The set of non-branching statements within the block.
    pub statements: Vec<hir::Statement>,
}

// Only print the number of statements in a block, makes debugging easier for now
impl fmt::Debug for Block {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}", self.statements)
    }
}

impl Block {
    /// Push a statement into the basic block
    pub fn push(&mut self, stmt: hir::Statement) {
        self.statements.push(stmt);
    }

    /// returns whether this block contains any statements
    pub fn is_empty(&self) -> bool {
        self.statements.is_empty()
    }

    pub fn exit_type(&self) -> BlockExit {
        match self.statements.last() {
            Some(stmt) => {
                use crate::hir::StatementKind;
                match stmt.kind {
                    StatementKind::BranchingCondition => BlockExit::BranchCondition,
                    StatementKind::LoopingCondition => BlockExit::LoopCondition,
                    StatementKind::Return => BlockExit::Return,
                    _ => BlockExit::Normal,
                }
            }
            None => BlockExit::Normal,
        }
    }
}

#[derive(Debug, Clone)]
pub enum BlockExit {
    // Placeholder, we don't know how the block exits yet
    Unknown,
    // The block forks based on a boolean condition. It will
    // have edges to other blocks for both `true` and `false`
    BranchCondition,
    LoopCondition,
    // The block ends with a return statement. It will have a
    // single edge to the special exit block.
    Return,
    // The block naturally flows to another block, e.g., when
    // an `if` block finishes executing.
    Normal,
}

#[derive(Debug, Clone)]
pub enum BasicBlock {
    /// A special block denoting the entry-point of the graph.
    Entry,
    /// A basic block.
    Block(Block),
    // Special block denoting the exit-point of the graph.
    Exit,
}

/// An edge between basic blocks. It answers the question
/// "how can control flow move between blocks?"
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ControlFlowEdge {
    /// Normal control flow is what occurs as code is executed
    /// top-down. For example, when an `if` without an `else` clause
    /// ends it naturally moves control flow back outside the block.
    Normal,
    /// When control flow depends on some boolean condition, this is
    /// the page taken when the condition evalutes to `true`.
    ConditionTrue,
    /// Same as ConditionTrue, but for `false`
    ConditionFalse,
    /// An explicit return statement, always goes straight to the exit block.
    Return,
}
