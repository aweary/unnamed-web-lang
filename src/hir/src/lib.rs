#![warn(clippy::all)]
#![allow(dead_code)]

#[cfg(test)]
extern crate quickcheck;
#[cfg(test)]
#[macro_use(quickcheck)]
extern crate quickcheck_macros;

// use insta;

pub mod hir;
pub mod lower;

mod control_flow_graph;
mod module_graph;
mod scope;

#[cfg(test)]

mod control_flow_graph_tests {
    use crate::control_flow_graph;
    use crate::lower;
    use insta::{assert_debug_snapshot, assert_snapshot};
    use parser::test_utils;
    use petgraph;
    use quickcheck;
    fn cfg_snapshot(body: &str) {
        let input = format!("fn foo() {{{}}}", body);
        let fn_ast = test_utils::parse_fn_def(&input);
        let fn_hir = lower::lower_fn(fn_ast);
        fn_hir.graph.dfs();
        assert_snapshot!(fn_hir.graph.graphviz_output());
    }

    fn cfg_from_code(input: &str) -> control_flow_graph::ControlFlowGraph {
        let fn_ast = test_utils::parse_fn_def(input);
        let fn_hir = lower::lower_fn(fn_ast);
        fn_hir.graph
    }
    fn cfg_from_block(block: &str) -> control_flow_graph::ControlFlowGraph {
        println!("BLOCK");
        let input = format!("fn foo() {}", block);
        println!("{}", input);
        let fn_ast = test_utils::parse_fn_def(&input);
        let fn_hir = lower::lower_fn(fn_ast);
        fn_hir.graph
    }

    #[quickcheck]
    fn is_graph(g: control_flow_graph::MockControlFlowCodeSnippetBuilder) -> bool {
        let code = g.code();
        let cfg = cfg_from_block(code);
        println!("\n\n\n{}\n\n\n", code);
        true
    }

    // Snapshot for a large and complex branching example
    #[test]
    fn kitchen_sink_branching() {
        cfg_snapshot(
            "
        let a = 1;
        let b = 2;
        if a == b {
            let c = 2;
            return c;
        } else if a > b {
            let d = 5;
            return d;
        } else if a < b {
            if a == b {
                return 5;
            };
            return 6;
        } else {
            let a = 1;
        };
        let d = 2;
        let c = 6;
        if d == c {
            let c = 5;
        } else {
            return 5;
        };
        if false {
            if true {
                let c = 5;
            } else {
                let d = 10;
            };
        };
        ",
        );
    }

    #[test]
    fn return_from_if_true() {
        cfg_snapshot(
            "
          if true {
              return 1;
          } else {
              let a = 1;
          };
          let a = 1;
        ",
        )
    }

    // Blocks with no branching.
    mod no_branching {
        use super::cfg_snapshot;

        #[test]
        fn empty() {
            cfg_snapshot("");
        }

        #[test]
        fn single() {
            cfg_snapshot(
                "
                let a = 1;
              ",
            );
        }
        #[test]
        fn double() {
            cfg_snapshot(
                "
                let a = 1;
                let b = 2;
              ",
            );
        }

        #[test]
        fn many() {
            cfg_snapshot(
                "
                let a = 1;
                let b = 2;
                let c = 3;
                let d = 4;
                let e = 5;
                let f = [6, 7, 8];
                let g = { nine: 10 };
              ",
            );
        }
    }

    mod if_branching {
        use super::*;
        #[test]
        fn basic() {
            cfg_snapshot(
                "
                if true {
                    let a = 1;
                };
                ",
            );
        }

        #[test]
        fn leading_single_stmt_basic() {
            cfg_snapshot(
                "
                let a = 1;
                if true {
                    let a = 1;
                };
                ",
            );
        }
        #[test]
        fn leading_multi_stmt_basic() {
            cfg_snapshot(
                "
                let a = 1;
                let b = 2;
                let c = 3;
                if true {
                    let a = 1;
                };
                ",
            );
        }
        #[test]
        fn trailing_single_stmt_basic() {
            cfg_snapshot(
                "
                if true {
                    let a = 1;
                };
                let c = 3;
                ",
            );
        }
        #[test]
        fn trailing_multi_stmt_basic() {
            cfg_snapshot(
                "
                if true {
                    let a = 1;
                };
                let c = 3;
                let d = 4;
                ",
            );
        }
    }
}
