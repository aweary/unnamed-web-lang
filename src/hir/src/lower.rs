use crate::control_flow_graph::{Block, BlockExit, BlockIndex, ControlFlowEdge, ControlFlowGraph};
use crate::hir;

use fxhash::FxHashSet;
use syntax::ast;



macro_rules! branch_block_metadata {
    ($block_indicies: ident, $root_block_indicies: ident, $cfg: ident, $partial_edges: ident) => {
        // It's possible that the then block is empty. If we were to compress
        // the CFG then this node would just be ignored, but we want to
        // potentially warn about empty blocks later so we create an empty block
        // for now.
        if $block_indicies.is_empty() {
            let empty_then_block = Block::default();
            let empty_block_index = $cfg.add_block(empty_then_block);
            $root_block_indicies.push(empty_block_index);
            let partial_edge = PartialFromEdge(empty_block_index, ControlFlowEdge::Normal);
            $partial_edges.insert(partial_edge);
            (empty_block_index, empty_block_index, false)
        } else {
            // Add the block indicies from the then block to the root set of
            // block indicies
            $root_block_indicies.extend($block_indicies.iter());
            let block_entry_node = *$block_indicies.first().unwrap();
            let block_exit_node = *$block_indicies.last().unwrap();
            let block_has_early_return = $cfg.block_has_early_exit(block_exit_node);
            // If this block exits early, add an edge to the root CFG's exit node.
            if block_has_early_return {
                $cfg.add_edge_to_exit(block_exit_node, ControlFlowEdge::Return);
            } else {
                // If the block doesn't exit early, queue up a partial edge, as control
                // will need to continue with whatever node comes after this branching
                // sub-graph
                let partial_edge = PartialFromEdge(block_exit_node, ControlFlowEdge::Normal);
                $partial_edges.insert(partial_edge);
            };
            (block_entry_node, block_exit_node, block_has_early_return)
        }
    };
}

pub fn lower_module(module: ast::Mod) {
    for item in module.items {
        if let ast::ItemKind::Fn(fn_def) = item.kind {
            lower_fn(fn_def);
            // ...
        };
    }
    // println!("Lowering module!");
}

pub(crate) fn lower_fn(fn_def: ast::FnDef) -> hir::Function {
    // The control flow graph for this function
    let mut cfg = ControlFlowGraph::default();
    // Initial function definition
    let fn_body_graph = lower_block(&*fn_def.body, &mut cfg);
    // Add an edge from the last item to the exit node
    if let Some(fn_body_graph_exit_node) = fn_body_graph.last() {
        if !cfg.block_has_early_exit(*fn_body_graph_exit_node) {
            cfg.add_edge_to_exit(*fn_body_graph_exit_node, ControlFlowEdge::Normal);
        }
    }
    // Any queued edges will point to the exit node, as we're done with the graph
    cfg.flush_edge_queue_to_exit_block();
    // cfg.print();
    hir::Function {
        span: fn_def.span,
        name: fn_def.name,
        graph: cfg,
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct PartialFromEdge(BlockIndex, ControlFlowEdge);
struct PartialToEdge(BlockIndex, ControlFlowEdge);

/// Computes a control flow graph for an if statement, This will be
/// treated as a sub-graph of the containing function's CFG, so it allocates
/// nodes in the outer CFG and returns the metadata needed to created the
/// required edges for the outer graph.
fn control_flow_graph_for_if_statement(
    ast::IfExpr {
        block: ref if_block,
        ref alt,
        ref span,
        ..
    }: &ast::IfExpr,
    cfg: &mut ControlFlowGraph,
    mut entry_block: Block,
) -> (Vec<BlockIndex>, FxHashSet<PartialFromEdge>) {
    // The set of partial edges that should be connected with the next node,
    // after this sub-graph is computed.
    let mut partial_edges_to_next_node = FxHashSet::default();
    let mut block_indicies: Vec<BlockIndex> = vec![];

    // Add the condition to the entry block
    entry_block.push(hir::Statement {
        kind: hir::StatementKind::BranchingCondition,
        span: *span,
    });

    // Add the entry block to the graph.
    // TODO should we just return a queue of statements instead of trying
    // to merge two nodes?
    let entry_block_index = cfg.add_block(entry_block);
    block_indicies.push(entry_block_index);

    let then_block_indicies = lower_block(if_block, cfg);

    let (then_block_entry_node, then_block_exit_node, then_block_has_early_return) = branch_block_metadata!(
        then_block_indicies,
        block_indicies,
        cfg,
        partial_edges_to_next_node
    );

    // Add an edge for the then block
    cfg.add_edge(
        entry_block_index,
        then_block_entry_node,
        ControlFlowEdge::ConditionTrue,
    );

    // Now we need to check for alternative branching for this if statement.
    // This is where things get tricky, as there could be an arbitrary number
    // of alternative branches if this is a series of else-if conditions.
    if let Some(alt) = alt {
        match alt {
            ast::Else::Block(ref block) => {
                let else_block_indicies = lower_block(block, cfg);

                // TODO dedupe with above
                let (else_block_entry_node, else_block_exit_node, else_block_has_early_return) = branch_block_metadata!(
                    else_block_indicies,
                    block_indicies,
                    cfg,
                    partial_edges_to_next_node
                );
                // Add a branch from the entry block to the else block when the condition
                // is false
                cfg.add_edge(
                    entry_block_index,
                    else_block_entry_node,
                    ControlFlowEdge::ConditionFalse,
                );
            }
            ast::Else::If(ref else_if) => {
                // Create a new block for this if else check.
                let alt_entry_block = Block::default();
                // Recurisvely create a sub-graph for this else-if branch. This will handled
                // arbitrarily deep if-else branching.
                let (alt_block_indicies, partial_edges_to_next_node_for_alt) =
                    control_flow_graph_for_if_statement(else_if, cfg, alt_entry_block);
                // The partial edges for this sub-graph need to be included in the root set,
                // which is what will be processed outside of this sub-graph
                partial_edges_to_next_node.extend(partial_edges_to_next_node_for_alt.iter());
                // Get the branch metadata for this alternative block
                let (alt_block_entry_node, alt_block_exit_node, alt_block_has_early_return) = branch_block_metadata!(
                    alt_block_indicies,
                    block_indicies,
                    cfg,
                    partial_edges_to_next_node
                );
                // Add an edge from the entry block to the alt `then` block, which will execute
                // if the root branch condition is false.
                cfg.add_edge(
                    entry_block_index,
                    alt_block_entry_node,
                    ControlFlowEdge::ConditionFalse,
                );
            }
        }
    } else {
        // If there is no alternative, control flow should continue on
        // to the next node--assuming the then block doesn't have an early return.
        let partial_edge = PartialFromEdge(entry_block_index, ControlFlowEdge::ConditionFalse);
        partial_edges_to_next_node.insert(partial_edge);
        // ...
    }
    (block_indicies, partial_edges_to_next_node)
}

fn lower_block(body: &ast::Block, cfg: &mut ControlFlowGraph) -> Vec<BlockIndex> {
    // Start off by creating a basic block. It's possible this block might be empty,
    // if this syntactic block is empty.
    let mut block_indicies = vec![];
    let mut block = Block::default();
    for stmt in &body.stmts {
        use ast::StmtKind::*;
        match stmt.kind {
            Local(_) => {
                // HIR statement node
                let stmt = hir::Statement {
                    kind: hir::StatementKind::LocalDefinition,
                    span: stmt.span,
                };
                // Add it to the current basic block
                block.push(stmt);
                // ...
            }
            If(ref if_expr) => {
                let branching_block = block;
                block = Block::default();
                let (if_branch_indicies, partial_edges) =
                    control_flow_graph_for_if_statement(if_expr, cfg, branching_block);
                block_indicies.extend(if_branch_indicies.iter());
                for PartialFromEdge(block_index, edge) in partial_edges {
                    cfg.enqueue_edge(block_index, edge);
                    // ...
                }
            }
            Return(_) => {
                // A return statement is branching, so it closes it current basic block
                // HIR statement for branching condition
                let stmt = hir::Statement {
                    kind: hir::StatementKind::Return,
                    span: stmt.span,
                };
                // Add to the basic block
                block.push(stmt);
                // Finalize the block, add it to the graph
                let block_index = cfg.add_block(block);
                block_indicies.push(block_index);
                // We know this block will only have an edge to the exit node, so create that
                // here.
                cfg.add_edge_to_exit(block_index, ControlFlowEdge::Return);
                // Create a new block, preempting any other statements in this syntax block
                block = Block::default();
            }
            While(ref condition, ref loop_block) => {
                let mut partial_edges_to_next_node = FxHashSet::default();
                // Close the previous block.
                let block_before_loop = block;
                let block_before_loop_index = if block_before_loop.is_empty() {
                    match cfg.last_block() {
                        Some(block_index) => block_index,
                        None => cfg.entry_block
                    }
                } else {
                    let block_before_loop_index = cfg.add_block(block_before_loop);
                    block_indicies.push(block_before_loop_index);
                    block_before_loop_index
                };
                // Create a new one for later...
                block = Block::default();
                // Create a new one for the loop condition
                let mut loop_condition_block = Block::default();
                // Add the condition statement
                loop_condition_block.push(hir::Statement {
                    kind: hir::StatementKind::LoopingCondition,
                    span: condition.span,
                });
                let loop_condition_block_index = cfg.add_block(loop_condition_block);

                if !cfg.block_has_early_exit(block_before_loop_index) {
                    cfg.add_edge(
                        block_before_loop_index,
                        loop_condition_block_index,
                        ControlFlowEdge::Normal,
                    );
                }

                block_indicies.push(loop_condition_block_index);
                // Lower the syntactic block for the loop

                let loop_body_indicies = lower_block(loop_block, cfg);
                // println!("enqueued edge {:?}", cfg.edge_queue);
                // The loop body has been graphed, so any queued edges will point back
                // to the loop condition, as that's where normal control flow will continue
                cfg.flush_edge_queue(loop_condition_block_index);

                let (loop_body_entry_node, loop_body_exit_node, loop_body_has_early_return) = branch_block_metadata!(
                    loop_body_indicies,
                    block_indicies,
                    cfg,
                    partial_edges_to_next_node
                );

                println!("loop_block_entry_node {:?}", loop_body_indicies);

                // Create an edge between the loop condition and the loop body
                // for when the loop condition is true.
                cfg.add_edge(
                    loop_condition_block_index,
                    loop_body_entry_node,
                    ControlFlowEdge::ConditionTrue,
                );
                // Create an edge from the loop body exit node to the loop condition,
                // so the condition can be checked again.
                if !loop_body_has_early_return {
                    cfg.add_edge(
                        loop_body_exit_node,
                        loop_condition_block_index,
                        ControlFlowEdge::Normal,
                    );
                }

                // Enqueue an edge to whatever block follows the loop for when the condition
                // is false.
                cfg.enqueue_edge(loop_condition_block_index, ControlFlowEdge::ConditionFalse);
            }
            _ => (),
        }
        // ...
    }

    // We've reached the end of this syntactic block.

    if !block.is_empty() {
        // Empty blocks are hard to deal with. In some cases we want to ignore them. For example,
        // if a function is empty we don't need to add this block. But in other cases, we want the
        // node in the graph, like when we have an if/else block that contains no statements.
        let block_index = cfg.add_block(block);
        block_indicies.push(block_index);
    }
    block_indicies
    // We've processed all the statements in this block.
}
