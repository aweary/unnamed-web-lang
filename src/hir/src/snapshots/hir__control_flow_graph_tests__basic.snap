---
source: src/lib.rs
expression: fn_hir.graph.graphviz_output()
---
digraph {
    0 [label="Entry"]
    1 [label="Exit"]
    2 [label="Block([BranchingCondition])"]
    3 [label="Block([LocalDefinition])"]
    0 -> 2 [label="Normal"]
    2 -> 3 [label="ConditionTrue"]
    3 -> 1 [label="Normal"]
    2 -> 1 [label="ConditionFalse"]
}

